File: 402_continuous-subarray-sum/continuous-subarray-sum.cpp
/*
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/continuous-subarray-sum
@Language: C++
@Datetime: 16-07-04 15:57
*/

class Solution {
public:
    /**
     * @param A an integer array
     * @return  A list of integers includes the index of 
     *          the first number and the index of the last number
     */
    vector<int> continuousSubarraySum(vector<int>& A) {
        // Write your code here
        if (A.size() == 0)
            return vector<int>(2, 0);
        vector<int> ret(2, 0);
        int curSum = 0;
        int maxSum = A[0];
        int prev = 0;
        int pos = 0;
        for (auto n: A) {
            if (curSum + n < n) {
                prev = pos;
                curSum = n;
            } else 
                curSum += n;
            if (maxSum < curSum) {
                maxSum = curSum;
                ret[0] = prev;
                ret[1] = pos;
            }

            ++pos;
        }
        return ret;
    }
};

File: 402_continuous-subarray-sum/continuous-subarray-sum.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   ultimate010
@Problem:  http://www.lintcode.com/problem/continuous-subarray-sum
@Language: Python
@Datetime: 16-06-27 16:13
'''

class Solution:
    # @param {int[]} A an integer array
    # @return {int[]}  A list of integers includes the index of the 
    #                  first number and the index of the last number
    def continuousSubarraySum(self, A):
        # Write your code here
        if not A:
            return [0, 0]
        f = A[0]
        start = 0
        ret = [0,0]
        m = f
        for i in range(1, len(A)):
            if m < f:
                ret = [start, i - 1]
                m = f
            if f + A[i] > A[i]:
                #chose a
                f = f + A[i]
            else:
                start = i
                f = A[i]
        if m < f:
            ret = [start, i]
        return ret